### DISCUSSION TITLE
Supporting cp38 or cp310 issue

### DISCUSSION THREAD

**Discussion by llggdd** [delta-io/delta-rs#1351] [Ideas]
I am trying to use this package for Azure Automations.
But it only supports 3.8 or 3.10 with manylinux2014 or 2_17.
It seems you don't support 3.8 or above.

Do you have a plan ? or can you upgrade this?
I tried myself to build 3.8 or above but as a rookie developer, I failed to build wheel.

**Comment by roeap** [comment]
Hi @llggdd, thanks for reaching out.

we are currently testing the dalta-rs python bindings against python 3.7 - 3.10, however we may need to drop 3.7 support soon (it's approaching EOL as well). Could you give a little more context on the environment you are trying to install deltalake in? In principle we are releasing wheels for the major architectures and python versions, so local builds would ideally not be needed.

If we cannot get that to work, could you be a bit more specific about the build errors? maybe we can provide some pointers on how the get the build to succeed.

**Reply by llggdd** [reply]
If I don't have a specific build for Azure Automation I cannot run the package at all. They don't offer terminal on its service. 
This is what it says on the whl upload page:
"Importing a Python package may take several minutes.
For Python 3.10 (preview) packages, only wheel files targeting cp310 Linux OS are currently supported. [Learn more](https://learn.microsoft.com/azure/automation/python-3-packages)
For Python 3.8 packages use wheel files targeting cp38-amd64, and for Python 2.7.12 use cp27-amd64."

**Reply by llggdd** [reply]
[
![image](https://github.com/delta-io/delta-rs/assets/54300720/3162df6f-5854-4676-bc81-e2bfc8807ffc)
](url)

**Reply by roeap** [reply]
Interesting, never worked with azure automation, but it seems it does break backwards compatibility. Generally our published wheels can be installed in python 3.7+ environments. It seems though that for some reason azure automation has an environment that cannot support wheels that also support 3.7.

Good news is we will be dropping 3.7 support fairly soon, as it goes EOL, at which point we would be publishing cp38 wheels.

I guess we could look into also publishing more wheel variants but contributor bandwidth is quite limited. 

Withtout having tried it, I guess you could build your own wheels as described in the python [readme](https://github.com/delta-io/delta-rs/blob/main/python/README.md). Before building I think you would have to change this [line](https://github.com/delta-io/delta-rs/blob/6b6911bc43ee5cee77ed8d3e267c4f44844fa5bb/python/Cargo.toml#LL37C1-L37C53) in the Cargo.toml to something lilke.

```yaml
[dependencies.pyo3]
version = "0.18"
features = ["extension-module", "abi3", "abi3-py38"]
```

**Reply by llggdd** [reply]
I tried myself to build own wheel, I failed. I was only successful on building a wheel for macOS. That's why I posted on this to get some help.  :)

**Comment by wjones127** [comment]
What do you mean we don't support python 3.8? What happens when you `pip install deltalake`? It installs the wheels fine in my test environment with Python 3.10.

**Reply by llggdd** [reply]
yes I also developed my code with Python 3.10, but try Azure Automation. You'll know in a minute. They only support .whl installation, not pip install. and they describe specific version of .whl that can be installed.

**Reply by llggdd** [reply]
[
![image](https://github.com/delta-io/delta-rs/assets/54300720/89f2678f-60bc-4ee5-be9d-1ac05e18166b)
](url)
you should install package one by one, with uploading whl.
